### ISSUE TITLE
Performance Issue in `box_collider::_stab`

### ISSUE THREAD

**Issue by trilean** [OpenXRay/xray-16#731]
The issue first appeared on the Skadovsk: when stashing a few things on the personal box there, the framerate dropped to around 10 FPS. This problem didn't exist at the beginning of the game, and it wasn't the first time I stashed something there. Removing all items from the stash didn't solve the problem.
From that point on, whenever the camera is pointed towards the Skadovsk, the framerate drops. The ship doesn't actually have to be visible for this to happen: When the player looks towards it, but the Skadovsk is obscured by a hill this will still happen.

Anyway, I didn't debug the issue at the time. But now I'm at the Jupiter Underground, and the problem returned much more severely: In the main hall, I dropped a few items because I was overloaded. Immediately the framerate dropped to around 1.4 FPS. Picking up the dropped items again didn't solve the issue. So I recompiled OpenXRay with the Google Perftools and did a CPU Profile. It immediately becomes clear where the problem is:

![pprof](https://user-images.githubusercontent.com/9497724/109153248-8e229c80-7764-11eb-8148-214d7801ba58.png)

This is for the problem in the Jupiter Underground. I started the game, loaded the savegame, turned around once and exited.

Reverting a61b93e151742baa39da1f3dd299e0e190b17908 did not solve the problem.

I'm running Gentoo Linux (x86_64) with Mesa 20.3.4 and X11. The code is from the `master` branch, though `dev` doesn't have any commits over `master` that change `src/xrCDB/xrCDB_box.cpp` (where `box_collider::_stab` is defined). FPS were measured using the Gallium HUD (built into Mesa).

**Comment by Xottab-DUTY** [comment]
@vTurbine, take a look on the graph. Any thoughts?

**Comment by eagleivg** [comment]
@trilean can you make attach save file with problem items? It would be ideal if you could find out which item causes this effect by being thrown/placed in a personal box.

**Comment by trilean** [comment]
@eagleivg sure here you go:

[test.scop.gz](https://github.com/OpenXRay/xray-16/files/6043258/test.scop.gz)

The items that I dropped in the Jupiter Underground when the problem appeared were either a bunch of Cossacks Vodkas or some .45 ACP FMJ rounds (the regular version, not the Hydroshocks). Unfortunately I don't remember which item i stashed on the Skadovsk.

**Comment by trilean** [comment]
To me it looks like `AABBNoLeafNode` (which should probably be a binary tree) forms a circle somehow and the recursion in `_stab` never ends. Can this possibly be caused by items being too close together?

**Comment by vTurbine** [comment]
@trilean thanks for such deep investigation.
At first glance I can guess that It’s a sync issue between renderer (right tree, very first branch with ‘ray_query’) and physics update (left tree, ‘box_query’ which is happening on a separate thread).
The renderer is likely sorting dynamic geometry (see ‘spatial_updatesector’ in ‘render_main’) and accessing the same spatial structure which is used by collision code.
If you check ‘box_query’ and ‘ray_query’ implementations you will see a ‘MODEL::synchronize()’ call which is actually a critical section.

**Comment by trilean** [comment]
So how do you propose we go about fixing this? I looked through the code, especially at the places you mentioned (`ray_query`, `box_query` and the locks they're using) but couldn't find anything obviously wrong.

**Comment by vTurbine** [comment]
I'm trying to figure it now, but have no luck with repro on your save file. Have found piles of vodka and ammo -- near this place FPS is stable 60 and never drops even when items were picked up and dropped again.
I'm on Windows, so your issue may appear on Linux builds only or requires a specific game configuration. @eagleivg have you managed to reproduce this?

**Comment by eagleivg** [comment]
Yes, this bug is only on linux, although until now it was not known what exactly causes it. With debugging I managed to find out that the problem is somewhere in CRender::detectSector() (or deeper), but I couldn't find where exactly. Another related fact is that most of the game saves where this problem was played now won't load because the R_ASSERT2(_valid(bi.mTransform), "anim kils bone matrix") check is on;
In general, the problem is waiting for its explorers:)

**Comment by eagleivg** [comment]
Thank to ⱱ2ⱱ3ⱱ4, found the most likely cause of the drop in fps - it is NaN in the float calculations. It is indirectly confirmed by the triggering of _valid. But the source of the problem is also unclear, now at least we can see that it is some FPE in the code generated by GCC. Should try this: https://linux.die.net/man/3/feenableexcept for debug/
